tooling/nargo_cli/src/cli/interpret_cmd.rs

```diff
@@ -0,0 +1,225 @@
+use noirc_frontend::monomorphization::ast::Program;
+use noirc_frontend::monomorphization::{monomorphize, monomorphize_debug};
+
+use crate::errors::CliError;
+
+use super::compile_cmd::parse_workspace;
+use super::{LockType, PackageOptions, WorkspaceCommand};
+
+/// Compile the program and interpret the SSA after each pass,
+/// printing the results to the console.
+#[derive(Debug, Clone, Args)]
+pub(super) struct InterpretCommand {
+    #[clap(flatten)]
+    pub(super) package_options: PackageOptions,
+
+    #[clap(flatten)]
+    pub(super) compile_options: CompileOptions,
+
+    /// The name of the TOML file which contains the ABI encoded inputs.
+    #[clap(long, short, default_value = PROVER_INPUT_FILE)]
+    prover_name: String,
+
+    /// The name of the SSA passes we want to interpret the results at.
+    ///
+    /// When nothing is specified, the SSA is interpreted after all passes.
+    #[clap(long)]
+    ssa_pass: Vec<String>,
+}
+
+impl WorkspaceCommand for InterpretCommand {
+    fn package_selection(&self) -> PackageSelection {
+        self.package_options.package_selection()
+    }
+
+    fn lock_type(&self) -> LockType {
+        // Does not write any artifacts.
+        LockType::Shared
+    }
+}
+
+pub(crate) fn run(args: InterpretCommand, workspace: Workspace) -> Result<(), CliError> {
+    let (file_manager, parsed_files) = parse_workspace(&workspace, None);
+    let binary_packages = workspace.into_iter().filter(|package| package.is_binary());
+
+    let ssa_options = to_ssa_options(&args.compile_options);
+    let ssa_passes = primary_passes(&ssa_options);
+
+    for package in binary_packages {
+        // Compile into monomorphized AST
+        let program_result = compile_into_program(
+            &file_manager,
+            &parsed_files,
+            &workspace,
+            package,
+            &args.compile_options,
+        );
+
+        // Report warnings and get the AST, or exit if the compilation failed.
+        let program = report_errors(
+            program_result,
+            &file_manager,
+            args.compile_options.deny_warnings,
+            args.compile_options.silence_warnings,
+        )?;
+
+        // Parse the inputs and convert them to what the SSA interpreter expects.
+        let abi = noir_ast_fuzzer::program_abi(&program);
+        let prover_file = package.root_dir.join(&args.prover_name).with_extension("toml");
+        let (prover_input, _) =
+            noir_artifact_cli::fs::inputs::read_inputs_from_file(&prover_file, &abi)?;
+        let ssa_input = noir_ast_fuzzer::input_values_to_ssa(&abi, &prover_input);
+
+        // Generate the initial SSA.
+        let mut ssa = generate_ssa(program)
+            .map_err(|e| CliError::Generic(format!("failed to generate SSA: {e}")))?;
+
+        print_and_interpret_ssa(&ssa_options, &args.ssa_pass, &mut ssa, "Initial SSA", &ssa_input);
+
+        // Run all SSA passes that match the filter, printing the result after each step.
```

Comment:
Shouldn't all passes run, but we'd only print the ones that were passed to the command line?

Reply:
There is a number of these filters now:
* `--show-ssa-pass` filters which SSA passes are printed
* `--skip-ssa-pass` is a list of SSA passes which are not actually run, which is something Michael found useful in testing
* `--ssa-pass` of the `interpreter` command lists the passes where we execute the interpreter. 

This comment is a bit confusing, I'll rephrase it. I meant that run the ones which aren't skipped, but altogether it's probably better not to say anything about filtering.